### PR TITLE
Unify various status errors into one HTTP error type.

### PR DESCRIPTION
**Pull request by Mathew-Estafanous**
issue #11393

At the moment, every http code has its own error type, which makes adding a new http code response more difficult than it should be.

This PR aims at unifying the various errors types (``BadRequestError``, ``ForbiddenError``) into one error type ``HTTPError``. This new type allows for much fewer error structs as well as the flexibility to returning new status codes.

**PR is not finished**

### CODE REVIEW

**Comment by kisunji** [comment]
Hmm that's weird. I think on Github the `go-tests` is stuck on 'In progress' but CircleCI reports all tests as passing. 

Are `TestAgent_*` tests still failing locally?

Unfortunately it's a huge package so we hardly run tests on them locally; it could be OS or resource-specific errors that we never caught

**Comment by Mathew-Estafanous** [comment]
Looks like ``TestAgent_Host`` is the only test that is consistently failing locally.
image

The other test cases passed when I ran them individually, which is a little weird.

**Comment by kisunji** [comment]
You're all green on the CircleCI side! Mark me as reviewer when you want a review (or ping me if you want a sanity check that CI is passing 🤣 )

**Comment by Mathew-Estafanous** [comment]
I see. Hopefully GitHub figure is out soon.

One thing before a review. I’m thinking of making a constructor for ``HTTPError`` that the handlers would use instead of initializing with a struct literal. My reasoning is that the wrapper makes the assumption that the status code and reason are filled out properly and the constructor would  ‘enforce’ this by requiring a status code and reason.

Constructor like this.
```go
func NewHTTPError(statusCode int, reason string) HTTPError {
	return HTTPError{statusCode, reason}
}
```

Then the handler would return the error like this.
```go
func (s *HTTPHandlers) SomeHandler(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
     // some code here...
    return nil, NewHTTPError(http.StatusBadRequest, “You didn’t format it right")
}
```

This might be overkill, but thought I’d suggest in case. 🤷🏽‍♂️ 

If what I suggested above shouldn’t be implemented then I’m ready for review!

**Comment by Mathew-Estafanous** [comment]
Also, I don’t have permission to add reviewers so I’ll probably have to stick with pinging you.

**Comment by Mathew-Estafanous** [comment]
@kisunji I pushed the change. Looks like everything passed, so I'm ready for a review.

File: agent/connect_auth.go
Comment (kisunji) [inline agent/connect_auth.go:27]:
```suggestion
// If the request input is invalid the error returned will be a BadRequest HTTPError,
```

File: agent/txn_endpoint_test.go
Comment (kisunji) [inline agent/txn_endpoint_test.go:7]: nit-picky but can we move this import to the 2nd import group?

File: agent/txn_endpoint_test.go
Comment (kisunji) [inline agent/txn_endpoint_test.go:75]:
```suggestion
				t.Fatalf("excected HTTP error but got %v", err)
```

File: agent/txn_endpoint_test.go
Comment (kisunji) [inline agent/txn_endpoint_test.go:158]:
```suggestion
				t.Fatalf("excected HTTP error but got %v", err)
```

**Review by kisunji — APPROVED** [review]
This is awesome! A few typo fixes then we can merge

**Comment by kisunji** [comment]
Thanks for your contribution! It might take a bit of time to merge because we have a separate enterprise repo that I'll have to manually update

**Comment by kisunji** [comment]
@Mathew-Estafanous Are you able to view the running jobs on your fork from CircleCI? It looks like this PR is stuck on a CI job but I can't tell where.

**Comment by Mathew-Estafanous** [comment]
Hmm, no I don’t see it either. I only see that ``verify-ci`` has been blocked, but there is no details linking to more information about it.

Should I push an empty commit to try restarting things?

**Comment by kisunji** [comment]
Yeah try it to see if it will unblock! Otherwise I can open another PR and cherrypick all your commits. Sometimes our CI behaves strangely with PRs from forks

**Comment by kisunji** [comment]
Hmm I think CircleCI is being smart and refusing to rerun the test since there's no diff. I re-ran it manually but in any case I think this is fine merging as-is (since I'm confident it won't impact integration tests)

**Comment by Mathew-Estafanous** [comment]
Sounds good! Also, for future issues should I ping you or should I just go about making the PR, etc and wait for others to review?

**Comment by kisunji** [comment]
You are welcome to continue pinging me but I may take longer to respond depending on my workload. Our team meets weekly to review new or updated issues and PRs so you'll ideally get a response from someone within a week when you raise a PR or make a comment on an issue

**Comment by kisunji** [comment]
Just updating to say I'm working on the enterprise side of things! I can fix the intentions_endpoint.go here as well

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/656224.

**Pull request merged**
